### PR TITLE
feat(deployment): update and secure metacontroller

### LINKED ISSUE #5578: can metacontroller be upgraded to latest version from V0.3.0
/kind question
/kind/question

can metacontroller be upgraded to latest version from V0.3.0
from v1.2.0 metacontroller have support for distroless image .

### PR DESCRIPTION
**Pull request by juliusvonkohout**
@Bobgy @zijianjoy @orugantichetan
 Fixes #5578

I
*  upgraded to 2.0.4
*  removed insane permissions (e.g. privileged root container)
*  removed insane cpu and memory ressources (4 cores 4 GB)
*  restricted the namespaces to kubeflow profile namespaces
*  Reduced to a sane amount of logging
*  Omitted networkattachmentdefinitions for rootless istio-cni on Openshift and i did not update the profile-controller accordingly. I can still add them if someone is interested. This would also add something to pipelines-profile-controller sync.py https://istio.io/latest/docs/setup/platform-setup/openshift/#additional-requirements-for-the-application-namespace. I can also create a second pull request for that. In the same pull request i could also add the poddefault for authentication with serviceaccounts https://github.com/kubeflow/pipelines/issues/5138#issuecomment-911399642 @bartgras

Sadly it still needs cluster-admin rights, but maybe someone has an idea on how to restrict it even further.

### CODE REVIEW

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml [hunk 1]
```diff
@@ -18,26 +18,28 @@ spec:
...
+            limits:
+              cpu: '1'
+              memory: 1Gi
+            requests:
+              cpu: 100m
+              memory: 256Mi
+          command:
+            - /usr/bin/metacontroller
+            - --zap-log-level=4
+            - '--discovery-interval=3600s' # less insane than 10 seconds
```
Comment (Bobgy) [inline manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:31]: What does this arg do? If a new namespace is created can required resources get created right away?

**Review by Bobgy — COMMENTED** [review]
/lgtm
/assign @zijianjoy
For final approval

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml [hunk 2]
```diff
@@ -18,26 +18,28 @@ spec:
...
+          securityContext:
+            capabilities:
+              drop:
+              - ALL
+            readOnlyRootFilesystem: true
+            runAsNonRoot: true
+            runAsUser: 1000
```
Comment (zijianjoy) [inline manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:38]: Any reason for the security context setup here?

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml [hunk 3]
```diff
@@ -18,26 +18,28 @@ spec:
...
-      - command:
-        - /usr/bin/metacontroller
-        - --logtostderr
-        - -v=4
-        - --discovery-interval=20s
-        image: metacontroller/metacontroller:v0.3.0
-        imagePullPolicy: Always
-        name: metacontroller
-        ports:
-        - containerPort: 2345
```
Comment (zijianjoy) [inline manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:30]: Why do we omit the port here?

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml (hunk 1 above)
Comment (juliusvonkohout) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:31]: Yes the resources are created right away in my installation.

From the documentation  https://metacontroller.github.io/metacontroller/guide/configuration.html?highlight=discovery-interval#command-line-flags

```
--discovery-interval | How often to refresh discovery cache to pick up newly-installed resources (e.g. --discovery-interval=10s). I think that means CRDs. A low value creates too much log spam and on start the cache is refreshed anyway. I am also fine with 10 minutes if that is what you want.

--zap-log-level | ZapGws  Level to configure the verbosity of logging. Can be one of ‘debug’,  ‘info’, ‘error’, or any integer value > 0 which corresponds to custom  debug levels of increasing verbosity(e.g. --zap-log-level=5).  Level 4 logs Metacontroller's interaction with the API server. Levels 5  and up additionally log details of Metacontroller's invocation of  lambda hooks. See the troubleshooting guide for more.
```

What you are worrying about is https://metacontroller.github.io/metacontroller/api/compositecontroller.html?highlight=discovery%20cache#resync-period but even that can be set to a high value in the compositecontroller "Sometimes you may want to sync periodically even if nothing has changed in the Kubernetes API objects"

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml (hunk 2 above)
Comment (juliusvonkohout) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:38]: yes, we obviously do want to run metacontroller with the smallest amount of permissions so an unpriviledged non-root. I can remove runAsUser: 1000 if you think that it is too much.

Comment (juliusvonkohout) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:38]: As you can see in the full source code it had a securitycontext before and was a privileged root container before.

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml (hunk 3 above)
Comment (juliusvonkohout) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:30]: The question should be rather: Why was the port mentioned before? It is not in upstream https://github.com/metacontroller/metacontroller/blob/master/manifests/production/metacontroller.yaml and i do not see why the port belongs there. I think this is a relict of the old 0.3 version

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml (hunk 2 above)
Comment (zijianjoy) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:38]: Got it, thank you for the explanation, we can keep runAsUser and we can come back if encountering any issue.

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml (hunk 3 above)
Comment (zijianjoy) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:30]: Sounds good to me.

**Comment by zijianjoy** [comment]
/approve

Thank you @juliusvonkohout

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml (hunk 1 above)
Comment (juliusvonkohout) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:31]: > /approve
> 
> Thank you @juliusvonkohout

Thank you too.

"To complete the pull request process, please ask for approval from bobgy after the PR has been reviewed." So i think @zijianjoy approval is not enough @bobgy

**Comment by juliusvonkohout** [comment]
> /approve
> 
> Thank you @juliusvonkohout

> /approve
> 
> Thank you @juliusvonkohout

Thank you too.

"To complete the pull request process, please ask for approval from bobgy after the PR has been reviewed." So i think @zijianjoy approval is not enough @bobgy

File: manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml (hunk 1 above)
Comment (juliusvonkohout) [reply manifests/kustomize/third-party/metacontroller/base/stateful-set.yaml:31]: Is this detailed enough, or should i test something for you ?

**Comment by Bobgy** [comment]
/lgtm
/approve
Thank you for the great contribution!

**Comment by sebastien-prudhomme** [comment]
@juliusvonkohout are you sure about the way "restricted the namespaces to kubeflow profile namespaces" is done?

It seems that here is no labelSelector field in parentResource field of CompositeController.

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout are you sure about the way "restricted the namespaces to kubeflow profile namespaces" is done?
> 
> It seems that here is no labelSelector field in parentResource field of CompositeController.
> 

I am not sure anymore.
Please try to move the labelselector one level higher and report back whether it makes a difference.

According to https://metacontroller.github.io/metacontroller/api/compositecontroller.html#label-selector you might have to alter sync.py to to add the labels to the child resources

Please also test profile/namespace deletion and creation.

**Comment by sebastien-prudhomme** [comment]
@juliusvonkohout the error I got with this commit is with K8S itself which does'nt want to create the CompositeController resource because `labelSelector` is an unknown field.

As I'm not using a standard environment, can you just confirm it works on your side?

**Comment by juliusvonkohout** [comment]
Correction: it seems to be fine with the label selector. It is the same on my clusters. Please delete your cluster, use a proper kubernetes 1.19-1.21 and try again.

**Comment by sebastien-prudhomme** [comment]
@juliusvonkohout found the problem: kubeflow-pipelines is not using the official CRD for CompositeController and I'm using the official one which has not the `preserveUnknownFields: true` option.

**Comment by juliusvonkohout** [comment]
@sebastien-prudhomme  then please create a new issue and pull request such that it is fixed long term for every one.

**Pull request merged**